---
title: How to convert GTEx v8 model to hg19 based on UK Biobank SNP set mapping
author: Sabrina Mi
date: '2020-11-23'
slug: how-to-convert-gtex-v8-model-to-hg19-based-on-uk-biobank-snp-set-mapping
categories: []
tags:
  - how to
  - prediction model
description: ''
topics: []
---

<link href="index_files/anchor-sections/anchor-sections.css" rel="stylesheet" />
<script src="index_files/anchor-sections/anchor-sections.js"></script>


<p><code>/gpfs/data/im-lab/nas40t2/Data/References/mappings/UKB2GTEx_mapping.txt.gz</code> contains information for variants in UK Biobank genotypes. The columns are</p>
<pre class="bash"><code>variant chromosome  position    non_effect_allele   effect_allele   rsid    zscore  panel_variant_id
1:692794:CA:C   chr1    757414  CA  C   1:692794_CA_C   1   NA
1:693731:A:G    chr1    758351  A   G   rs12238997  1   chr1_758351_A_G_b38
1:707522:G:C    chr1    772142  G   C   rs371890604 1   chr1_772142_G_C_b38
1:717587:G:A    chr1    782207  G   A   rs144155419 1   chr1_782207_G_A_b38</code></pre>
<p>MetaXcan prediction models has its own format to identify variants, so a mapping file like <code>UKB2GTEx_mapping.txt.gz</code> allows us to convert the variants in prediction models to the corresponding ID used in UK Biobank data. The variant column has the variant’s ID in the UK Biobank data and the panel_variant_id column has it’s ID used in prediction models, specifically the GTEx V8 models.</p>
<p><code>/gpfs/data/im-lab/nas40t2/Data/References/mappings/map_GTEx_v8_models_to_UKB_SNPset.tar</code> has a mashr model trained on GTEx V8 Whole Blood tissue along with its covariance matrix. They have variant ids in the panel_variant_id chr_pos_ref_alt_b38 format, which we want to swap to the UK Biobank format. Unpack it:</p>
<pre class="bash"><code>tar -xvf /gpfs/data/im-lab/nas40t2/Data/References/mappings/map_GTEx_v8_models_to_UKB_SNPset.tar</code></pre>
<p>The prediction models are SQLite databases, which can be queried in R with the <a href="https://www.datacamp.com/community/tutorials/sqlite-in-r">RSQLite</a> package.</p>
<p>Start by reading the prediction model into R.</p>
<pre class="r"><code>library(data.table)
mapping &lt;- fread(&quot;/gpfs/data/im-lab/nas40t2/Data/References/mappings/UKB2GTEx_mapping.txt.gz&quot;)</code></pre>
<p>Next, load the prediction model.</p>
<pre class="r"><code>library(RSQLite)
conn &lt;- dbConnect(RSQLite::SQLite(), &quot;/gpfs/data/im-lab/nas40t2/Data/References/mappings/map_GTEx_v8_models_to_UKB_SNPset/mashr_Whole_Blood.db&quot;)
dbListTables(conn)
weights &lt;- dbGetQuery(conn, &#39;SELECT * FROM weights&#39;)
extra &lt;- dbGetQuery(conn, &#39;SELECT * FROM extra&#39;)
dbDisconnect(conn)</code></pre>
<p>The mashr prediction model has two tables weights and extra, and the dbGetQuery calls pull all entries from the weights and extra table.</p>
<p>We only need the two columns, variant and panel_variant_id. Then we inner join the mapping with the variants in the weights table.</p>
<pre class="r"><code>library(tidyverse)
mapping &lt;- select(mapping, variant, panel_variant_id)
mapped_weights &lt;- inner_join(weights, mapping, by=c(&quot;varID&quot; = &quot;panel_variant_id&quot;))
mapped_weights &lt;- mapped_weights %&gt;% mutate(varID = variant) %&gt;% select(-variant)</code></pre>
<p>Update the n.snps.in.model column in the extra table:</p>
<pre class="r"><code>n.snps &lt;- mapped_weights %&gt;% group_by(gene) %&gt;% summarise(n.snps.in.model = n())
updated_extra &lt;- inner_join(n.snps, extra %&gt;% select(-n.snps.in.model), by=&quot;gene&quot;)</code></pre>
<p>Write the new tables to a new RSQLite database file.</p>
<pre class="r"><code>conn &lt;- dbConnect(RSQLite::SQLite(), &quot;/gpfs/data/im-lab/nas40t2/Data/References/mappings/map_GTEx_v8_models_to_UKB_SNPset/mashr_Whole_Blood_UKB.db&quot;)
dbWriteTable(conn, &quot;weights&quot;, mapped_weights)
dbWriteTable(conn, &quot;extra&quot;, updated_extra)
dbDisconnect(conn)</code></pre>
<p>Similarly, we inner join to convert both RSID1 and RSID2 columns in <code>/gpfs/data/im-lab/nas40t2/Data/References/mappings/map_GTEx_v8_models_to_UKB_SNPset/map_GTEx_v8_models_to_UKB_SNPset/mashr_Whole_Blood.txt.gz</code></p>
<pre class="r"><code>covariance &lt;- fread(&quot;/gpfs/data/im-lab/nas40t2/Data/References/mappings/map_GTEx_v8_models_to_UKB_SNPset/map_GTEx_v8_models_to_UKB_SNPset/mashr_Whole_Blood.txt.gz&quot;)
covariance &lt;- inner_join(covariance, merged, by=c(&quot;RSID1&quot; = &quot;varID&quot;)) %&gt;% mutate(RSID1 = variant) %&gt;% select(-variant)
covariance &lt;- inner_join(covariance, merged, by=c(&quot;RSID2&quot; = &quot;varID&quot;)) %&gt;% mutate(RSID2 = variant) %&gt;% select(-variant)
covariance &lt;- covariance %&gt;% select(GENE, variant.x, variant.y, VALUE) %&gt;% rename(RSID1 = variant.x, RSID2 = variant.y) 
write.table(covariance, &quot;/gpfs/data/im-lab/nas40t2/Data/References/mappings/map_GTEx_v8_models_to_UKB_SNPset/map_GTEx_v8_models_to_UKB_SNPset/mashr_Whole_Blood_UKB.txt&quot;)</code></pre>
